components/feature/autofill/src/main/java/mozilla/components/feature/autofill/structure/ParsedStructureBuilder.kt

```diff
@@ -0,0 +1,230 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.autofill.structure
+
+import android.os.Build
+import android.view.View
+import androidx.annotation.RequiresApi
+
+@RequiresApi(Build.VERSION_CODES.O)
+@Suppress("TooManyFunctions")
+internal class ParsedStructureBuilder<ViewNode, AutofillId>(
+    private val navigator: AutofillNodeNavigator<ViewNode, AutofillId>
+) {
+    fun build(): ParsedStructureData<AutofillId> {
+        val formNode = findFocusedForm()
+        val (usernameId, passwordId) = findAutofillIds(formNode)
+        val hostnameClue = usernameId ?: passwordId
+
+        return navigator.build(
+            usernameId,
+            passwordId,
+            getWebDomain(hostnameClue),
+            getPackageName(hostnameClue) ?: navigator.activityPackageName
+        )
+    }
+
+    private fun findFocusedForm(): ViewNode? {
+        val focusPath = findMatchedNodeAncestors {
+            navigator.isFocused(it)
+        }
+
+        return focusPath?.lastOrNull {
+            navigator.isHtmlForm(it)
+        }
+    }
+
+    private fun findAutofillIds(rootNode: ViewNode?): Pair<AutofillId?, AutofillId?> =
+        checkForAdjacentFields(rootNode) ?: getUsernameId(rootNode) to getPasswordId(rootNode)
+
+    private fun getUsernameId(rootNode: ViewNode?): AutofillId? {
+        // how do we localize the "email" and "username"?
```

Comment:
In theory i guess we can keep expanding the list below with commonly used terms, it may just work well enough.